specs/latest/2.0/index.html

```diff
@@ -1092,6 +1095,63 @@ <h2 class="no-toc">Table of contents</h2>
         it is the responsibility of the WebGL API to ensure that data are accessed consistently. This applies
         even if the buffer is currently bound to a transform feedback binding point.
       </dd>
+      <dt>
+        <p class="idl-code">Promise&lt;ArrayBuffer&gt; getBufferSubDataAsync(GLenum target, GLintptr offset, ArrayBuffer returnedData, unsigned long returnedOffset = 0, unsigned long returnedSize = 0)
+        </p>
+      </dt>
+      <dd>
+        Let <code>buffer</code> be the WebGLBuffer bound to the passed
+        target. Read <code>numBytes</code> bytes from <code>buffer</code> starting at byte
+        offset <code>offset</code> and write them to <code>returnedData</code> starting at byte
+        offset <code>returnedOffset</code>. <code>numBytes</code> is computed via the following
+        algorithm:
+
+        <ul>
+        <li> If <code>returnedSize</code> is zero, <code>numBytes = returnedData.byteLength - returnedOffset</code></li>
+        <li> Otherwise, <code>numBytes = returnedSize</code>.
+        </ul>
+
+        This command executes serially with respect to others in the command stream, but the data is
+        only written to <code>returnedData</code> when the promise is fulfilled; it is available in
+        the Promise's success callback, but not earlier.
+
+        <br><br>
+        If a failure occurs that prevents the promise from being fulfilled, such as:
+
+        <ul>
+          <li>no WebGLBuffer is bound to <code>target</code></li>
+          <li><code>offset</code> is less than zero</li>
+          <li><code>offset + numBytes</code> would extend beyond the end of <code>buffer</code></li>
+          <li><code>returnedOffset + numBytes</code> would extend beyond the end of <code>returnedData</code>
+          <li><code>target</code> is <code>TRANSFORM_FEEDBACK_BUFFER</code>, and any transform feedback object
+            is currently active</li>
+          <li>the context is lost before the Promise is fulfilled</li>
+        <ul>
+
+        <br><br>
+
+        then the returned Promise is rejected with an "InvalidStateError" DOMException, and no data
+        is written to <code>returnedData</code>.
+
+        <br><br>
+
+        Otherwise, the Promise is fulfilled when the data becomes available,
+        and <code>returnedData</code>, originally passed to <code>getBufferSubDataAsync</code>, is
+        passed to the success callback.
+
+        <br><br>
+        If the buffer is written and read sequentially by other operations
+        and <code>getBufferSubDataAsync</code>, it is the responsibility of the WebGL API to ensure
+        that data are accessed consistently. This applies even if the buffer is currently bound to a
+        transform feedback binding point.
+
```

Comment:
Yes, the WebGL implementation is responsible for ensuring that the returned data isn't visible until just before the promise is fulfilled. This does probably mean reading back into a separate buffer, and then copying it into this destination buffer on the main thread just before resolving the promise.

It's a good question about the guarantees if doing multiple asynchronous readbacks into the same buffer. For sanity's sake, I think that yes, the first Promise must be resolved before the second one. Do you foresee any problems with providing this guarantee?